File: src/modules/documents/document.service.ts
import { db } from "@/db";
import { documents } from "@/db/schema";
import { eq } from "drizzle-orm";
import {
  generateStorageKey,
  getUploadUrl,
  getDownloadUrl,
  deleteObject,
} from "./storage.adapter";

export const documentService = {
  async create(input: {
    decisionId: string;
    filename: string;
    originalFilename: string;
    mimeType: string;
    sizeBytes: number;
    storageKey: string;
    classification: string;
    uploadedBy: string;
  }) {
    const [doc] = await db
      .insert(documents)
      .values({
        decisionId: input.decisionId,
        filename: input.filename,
        originalFilename: input.originalFilename,
        mimeType: input.mimeType,
        sizeBytes: input.sizeBytes,
        storageKey: input.storageKey,
        classification: input.classification as "evidence" | "legal_opinion" | "correspondence" | "public_notice" | "internal_memo",
        uploadedBy: input.uploadedBy,
      })
      .returning();

    return doc;
  },

  async getByDecision(decisionId: string) {
    return db
      .select()
      .from(documents)
      .where(eq(documents.decisionId, decisionId));
  },

  async getById(id: string) {
    const [doc] = await db
      .select()
      .from(documents)
      .where(eq(documents.id, id))
      .limit(1);

    return doc ?? null;
  },

  async delete(id: string) {
    const doc = await this.getById(id);
    if (doc) {
      await deleteObject(doc.storageKey);
      await db.delete(documents).where(eq(documents.id, id));
    }
  },

  async getUploadUrl(
    decisionId: string,
    filename: string,
    contentType: string,
    classification: string,
    uploadedBy: string,
  ) {
    const storageKey = generateStorageKey(decisionId, filename);

    const doc = await this.create({
      decisionId,
      filename,
      originalFilename: filename,
      mimeType: contentType,
      sizeBytes: 0,
      storageKey,
      classification,
      uploadedBy,
    });

    const uploadUrl = await getUploadUrl(storageKey, contentType);

    return {
      uploadUrl,
      storageKey,
      documentId: doc.id,
    };
  },

  async confirmUpload(documentId: string, sizeBytes: number) {
    const [doc] = await db
      .update(documents)
      .set({ sizeBytes })
      .where(eq(documents.id, documentId))
      .returning();

    return doc;
  },

  async getDownloadUrl(storageKey: string) {
    return getDownloadUrl(storageKey);
  },
};

File: backend/src/DiscretionaryPowers.Infrastructure/Services/DocumentService.cs
using DiscretionaryPowers.Application.Common;
using DiscretionaryPowers.Application.DTOs.Documents;
using DiscretionaryPowers.Domain.Entities;
using DiscretionaryPowers.Domain.Enums;
using DiscretionaryPowers.Domain.Interfaces;
using DiscretionaryPowers.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace DiscretionaryPowers.Infrastructure.Services;

public class DocumentService(AppDbContext db, IStorageAdapter storage, ITenantService tenantService) : IDocumentService
{
    public async Task<ServiceResult<UploadUrlResponse>> GetUploadUrl(
        Guid decisionId, string filename, string contentType, string classification, Guid uploadedBy)
    {
        if (!Enum.TryParse<DocumentClassification>(classification, true, out var cls))
            return ServiceResult<UploadUrlResponse>.Fail("Invalid document classification.");

        var storageKey = $"decisions/{decisionId}/{Guid.NewGuid()}/{filename}";

        var document = new Document
        {
            Id = Guid.NewGuid(),
            DecisionId = decisionId,
            OrganizationId = tenantService.CurrentTenantId ?? Guid.Empty,
            Filename = filename,
            OriginalFilename = filename,
            MimeType = contentType,
            SizeBytes = 0,
            StorageKey = storageKey,
            Classification = cls,
            UploadedBy = uploadedBy,
            CreatedAt = DateTime.UtcNow,
        };

        db.Documents.Add(document);
        await db.SaveChangesAsync();

        var uploadUrl = await storage.GetUploadUrl(storageKey, contentType);

        return ServiceResult<UploadUrlResponse>.Ok(new UploadUrlResponse
        {
            UploadUrl = uploadUrl,
            StorageKey = storageKey,
            DocumentId = document.Id,
        });
    }

    public async Task<ServiceResult<DocumentResponse>> ConfirmUpload(Guid documentId, int sizeBytes)
    {
        var doc = await db.Documents.FindAsync(documentId);
        if (doc is null)
            return ServiceResult<DocumentResponse>.Fail("Document not found.");

        doc.SizeBytes = sizeBytes;
        await db.SaveChangesAsync();

        return ServiceResult<DocumentResponse>.Ok(MapToResponse(doc));
    }

    public async Task<List<DocumentResponse>> GetByDecision(Guid decisionId)
    {
        var docs = await db.Documents
            .AsNoTracking()
            .Where(d => d.DecisionId == decisionId)
            .OrderByDescending(d => d.CreatedAt)
            .ToListAsync();

        return docs.Select(MapToResponse).ToList();
    }

    public async Task<ServiceResult<object>> GetDownloadUrl(Guid documentId)
    {
        var doc = await db.Documents.AsNoTracking().FirstOrDefaultAsync(d => d.Id == documentId);
        if (doc is null)
            return ServiceResult<object>.Fail("Document not found.");

        var url = await storage.GetDownloadUrl(doc.StorageKey);
        return ServiceResult<object>.Ok(new { url, filename = doc.OriginalFilename });
    }

    public async Task<ServiceResult<bool>> Delete(Guid documentId)
    {
        var doc = await db.Documents.FindAsync(documentId);
        if (doc is null)
            return ServiceResult<bool>.Fail("Document not found.");

        await storage.DeleteObject(doc.StorageKey);
        db.Documents.Remove(doc);
        await db.SaveChangesAsync();

        return ServiceResult<bool>.Ok(true);
    }

    // IDocumentService interface methods
    async Task<Document> IDocumentService.Create(Guid decisionId, string filename, string originalFilename, string mimeType, int sizeBytes, string storageKey, DocumentClassification classification, Guid uploadedBy)
    {
        var doc = new Document
        {
            Id = Guid.NewGuid(),
            DecisionId = decisionId,
            OrganizationId = tenantService.CurrentTenantId ?? Guid.Empty,
            Filename = filename,
            OriginalFilename = originalFilename,
            MimeType = mimeType,
            SizeBytes = sizeBytes,
            StorageKey = storageKey,
            Classification = classification,
            UploadedBy = uploadedBy,
            CreatedAt = DateTime.UtcNow,
        };
        db.Documents.Add(doc);
        await db.SaveChangesAsync();
        return doc;
    }

    async Task<IReadOnlyList<Document>> IDocumentService.GetByDecision(Guid decisionId) =>
        await db.Documents.AsNoTracking().Where(d => d.DecisionId == decisionId).ToListAsync();

    async Task<Document?> IDocumentService.GetById(Guid id) =>
        await db.Documents.AsNoTracking().FirstOrDefaultAsync(d => d.Id == id);

    async Task IDocumentService.Delete(Guid id)
    {
        var doc = await db.Documents.FindAsync(id);
        if (doc is not null)
        {
            db.Documents.Remove(doc);
            await db.SaveChangesAsync();
        }
    }

    async Task<string> IDocumentService.GetUploadUrl(string key, string contentType) =>
        await storage.GetUploadUrl(key, contentType);

    async Task IDocumentService.ConfirmUpload(Guid documentId)
    {
        var doc = await db.Documents.FindAsync(documentId);
        if (doc is not null)
            await db.SaveChangesAsync();
    }

    private static DocumentResponse MapToResponse(Document d) => new()
    {
        Id = d.Id,
        DecisionId = d.DecisionId,
        Filename = d.Filename,
        OriginalFilename = d.OriginalFilename,
        MimeType = d.MimeType,
        SizeBytes = d.SizeBytes,
        StorageKey = d.StorageKey,
        Classification = d.Classification.ToString().ToLowerInvariant(),
        UploadedBy = d.UploadedBy,
        Version = d.Version,
        IsRedacted = d.IsRedacted,
        RedactionNotes = d.RedactionNotes,
        CreatedAt = d.CreatedAt,
    };
}
